pkg/controllers/infracluster/azure.go

```diff
@@ -0,0 +1,251 @@
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/yaml"
+)
+
+var (
+	errUnableToGetAzureMAPIProviderSpec = errors.New("unable to get Azure MAPI ProviderSpec")
+)
+
+const (
+	clusterSecretName               = "capz-manager-cluster-credential"    // #nosec G101
+	capzManagerBootstrapCredentials = "capz-manager-bootstrap-credentials" // #nosec G101
+)
+
+// ensureAzureCluster ensures the AzureCluster cluster object exists.
+//
+//nolint:funlen,cyclop
+func (r *InfraClusterController) ensureAzureCluster(ctx context.Context, log logr.Logger) (client.Object, error) {
+	// Get the Azure Bootstrap Credentials Secret.
+	// This is created by the Cluster Credential Operator and should always exist. We expect to always find it, if not, we error.
+	capzManagerBootstrapSecret := &corev1.Secret{}
+	if err := r.Get(ctx, client.ObjectKey{Namespace: defaultCAPINamespace, Name: capzManagerBootstrapCredentials}, capzManagerBootstrapSecret); err != nil {
+		if !cerrors.IsNotFound(err) {
+			return nil, fmt.Errorf("failed to get Azure Boostrap Credentials Secret: %w", err)
+		}
+	}
+
+	// Get the Azure Cluster Secret.
+	// CAPZ controllers expect a Secret Ref with a different data structure to what the secret we get from the Cluster Credential Operator above provides.
+	// That is why we are creating a new secret below to copy the values to.
+	clusterSecret := &corev1.Secret{}
+	if err := r.Get(ctx, client.ObjectKey{Namespace: defaultCAPINamespace, Name: clusterSecretName}, clusterSecret); err != nil {
+		if !cerrors.IsNotFound(err) {
+			return nil, fmt.Errorf("failed to get Azure Cluster Secret: %w", err)
+		}
+
+		// The Azure Cluster Secret does not exist, so it needs to be created.
+		if err := createNewAzureSecret(ctx, r.Client, capzManagerBootstrapSecret.Data["azure_client_secret"]); err != nil {
+			return nil, fmt.Errorf("failed to create Azure Cluster secret: %w", err)
+		}
+	}
+
+	apiURL, err := url.Parse(r.Infra.Status.APIServerInternalURL)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse apiUrl: %w", err)
+	}
+
+	port, err := strconv.ParseInt(apiURL.Port(), 10, 32)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse apiUrl port: %w", err)
+	}
+
+	if r.Infra.Status.PlatformStatus == nil {
+		return nil, fmt.Errorf("infrastructure PlatformStatus should not be nil: %w", err)
+	}
+
+	azureClientID, ok := capzManagerBootstrapSecret.Data["azure_client_id"]
+	if !ok {
+		return nil, fmt.Errorf("failed to get azureClientID: %w", err)
+	}
+
+	azureTenantID, ok := capzManagerBootstrapSecret.Data["azure_tenant_id"]
+	if !ok {
+		return nil, fmt.Errorf("failed to get azureTenantID: %w", err)
+	}
+
+	azureClusterIdentity := &azurev1.AzureClusterIdentity{}
+	// Get the Azure Cluster Identity.
+	if err := r.Get(ctx, client.ObjectKey{Namespace: defaultCAPINamespace, Name: r.Infra.Status.InfrastructureName}, azureClusterIdentity); err != nil && !cerrors.IsAlreadyExists(err) {
```

Comment:
Why would you have an `AlreadyExists` error on a `Get` call?

Reply:
I guess the intention here was to check if it already exists _with_ the contents we expect, and error if it doesn't, since we don't want to use a ClusterIdentity with the incorrect info.
@damdo does that match up with what our intentions should be?

Reply:
My point here is that you're doing a get call, not a create. You wouldn't get an already exists on a get call, ever.